sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -143,130 +141,29 @@ public TransportShardUpsertAction(ThreadPool threadPool,
-                false,
-                sourceToParse
-            );
-            if (indexResult.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                // Even though the primary waits on all nodes to ack the mapping changes to the master
-                // (see MappingUpdatedAction.updateMappingOnMaster) we still need to protect against missing mappings
-                // and wait for them. The reason is concurrent requests. Request r1 which has new field f triggers a
-                // mapping update. Assume that that update is first applied on the primary, and only later on the replica
-                // (it’s happening concurrently). Request r2, which now arrives on the primary and which also has the new
-                // field f might see the updated mapping (on the primary), and will therefore proceed to be replicated
-                // to the replica. When it arrives on the replica, there’s no guarantee that the replica has already
-                // applied the new mapping, so there is no other option than to wait.
-                throw new TransportReplicationAction.RetryOnReplicaException(indexShard.shardId(),
-                    "Mappings are not available on the replica yet, triggered update: " + indexResult.getRequiredMappingUpdate());
-            }
-            location = indexResult.getTranslogLocation();
-        }
-        return new WriteReplicaResult<>(request, location, null, indexShard, logger);
+        return new Context(valueValidation, insertSourceGen, updateSourceGen, returnValueGen, request.duplicateKeyAction());
     }
 
     @Nullable
-    private IndexItemResponse indexItem(ShardUpsertRequest request,
-                                        ShardUpsertRequest.Item item,
-                                        IndexShard indexShard,
-                                        @Nullable UpdateSourceGen updateSourceGen,
-                                        @Nullable InsertSourceGen insertSourceGen,
-                                        @Nullable ReturnValueGen returnValueGen) throws Exception {
+        TransportShardIndexAction.IndexItemResponse processItem(ShardUpsertRequest.Item item, Context context, IndexShard indexShard) throws Exception {
         VersionConflictEngineException lastException = null;
         boolean tryInsertFirst = item.insertValues() != null;
         boolean isRetry;
         for (int retryCount = 0; retryCount < MAX_RETRY_LIMIT; retryCount++) {
             try {
                 isRetry = retryCount > 0;
                 if (tryInsertFirst) {
-                    return insert(request, item, indexShard, isRetry, returnValueGen, insertSourceGen);
+                    return insert(item, indexShard, isRetry, context);
                 } else {
-                    return update(item, indexShard, isRetry, returnValueGen, updateSourceGen);
+                    return update(item, indexShard, isRetry, context);
```

Comment:
This is where gets ugly, `insert` and `update` are still duplicated, because the other implementation are depending on the actually type such as `ShardUpdateRequest.Item` etc. I could remove all types at all and just work with the values but this would feel i am working against the typesystem, other option could be push all values into one `Item` but then i could stick to the original implemenation.